### PR TITLE
gh-128013: fix data race in PyUnicode_AsUTF8AndSize on free-threading

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-128013

### CODE REVIEW

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -5823,7 +5823,11 @@ unicode_fill_utf8(PyObject *unicode)
 {
     /* the string cannot be ASCII, or PyUnicode_UTF8() would be set */
     assert(!PyUnicode_IS_ASCII(unicode));
-
+    int ret = 0;
+    Py_BEGIN_CRITICAL_SECTION(unicode);
```
Comment (vstinner) [inline Objects/unicodeobject.c]: Would it be possible to move the `assert(!PyUnicode_IS_ASCII(unicode));` assertion inside the critical section?

File: Lib/test/test_capi/test_unicode.py [hunk 2]
```diff
@@ -1005,6 +1005,23 @@ def test_asutf8(self):
         self.assertRaises(TypeError, unicode_asutf8, [], 0)
         # CRASHES unicode_asutf8(NULL, 0)
 
+    @unittest.skipIf(_testcapi is None, 'need _testcapi module')
+    @threading_helper.requires_working_threading()
+    def test_asutf8_race(self):
+        from _testcapi import unicode_asutf8
```
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py]: Please add a comment to explain the purpose of this test.

Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py]: No need to import _testcapi, it's already imported. Just use `unicode_asutf8 = _testcapi.unicode_asutf8`, no?

File: Objects/unicodeobject.c [hunk 3]
```diff
@@ -118,14 +118,14 @@ NOTE: In the interpreter's initialization phase, some globals are currently
     (assert(_PyUnicode_CHECK(op)),                      \
      PyUnicode_IS_COMPACT_ASCII(op) ?                   \
          ((char*)(_PyASCIIObject_CAST(op) + 1)) :       \
-         _PyUnicode_UTF8(op))
+        FT_ATOMIC_LOAD_PTR(_PyUnicode_UTF8(op)))
```
Comment (colesbury) [inline Objects/unicodeobject.c]: I don't think we should put the atomics operations in the macros -- it's too easy to hide data races that way.

`PyUnicode_UTF8()` functions as both a guard and as an accessor. We should split those up.

File: Objects/unicodeobject.c [hunk 4]
```diff
@@ -118,14 +118,14 @@ NOTE: In the interpreter's initialization phase, some globals are currently
...
 #define _PyUnicode_UTF8_LENGTH(op)                      \
     (_PyCompactUnicodeObject_CAST(op)->utf8_length)
 #define PyUnicode_UTF8_LENGTH(op)                       \
     (assert(_PyUnicode_CHECK(op)),                      \
      PyUnicode_IS_COMPACT_ASCII(op) ?                   \
          _PyASCIIObject_CAST(op)->length :              \
-         _PyUnicode_UTF8_LENGTH(op))
+         FT_ATOMIC_LOAD_SSIZE_RELAXED(_PyUnicode_UTF8_LENGTH(op)))
```
Comment (colesbury) [inline Objects/unicodeobject.c]: No atomic operation here.

**Review by colesbury — COMMENTED** [review]
I think there's still a bug where `PyUnicode_UTF8()` is checked outside the lock, but the condition may change once the lock is acquired (because some other thread filled in the `utf8` field).

I think we should refactor out the check into something like `unicode_ensure_utf8` that does the double-checked locking:

```c

static int
unicode_ensure_utf8(PyObject *unicode)
{
    int err = 0;
    if (PyUnicode_UTF8(unicode) == NULL) {
        Py_BEGIN_CRITICAL_SECTION(unicode);
        if (PyUnicode_UTF8(unicode) == NULL) {
            err = unicode_fill_utf8(unicode);
        }
        Py_END_CRITICAL_SECTION();
    }
    return err;
}
```

`unicode_fill_utf8` should assert that the critical section is held.

File: Objects/unicodeobject.c (hunk 4 above)
Comment (colesbury) [reply Objects/unicodeobject.c]: I think if we need an atomic operation here to silence some TSAN warning, it's a sign of a bug somewhere else.

The atomic synchronization should happen on just one field: `utf8`. If `utf8` is not-NULL, then `utf8_length` should be valid (and not require any atomics to read).

**Comment by vstinner** [comment]
I wrote PR gh-128061 "Convert unicodeobject.c macros to functions" to prepare the code for this change.

**Comment by kumaraditya303** [comment]
I have updated the PR to use the new static inline functions and it now uses acquire/release semantics for utf8 member. I have tested the reproducer from issue and now there aren't any data races AFAICS.

File: Objects/unicodeobject.c [hunk 5]
```diff
@@ -700,15 +700,15 @@ _PyUnicode_CheckConsistency(PyObject *op, int check_content)
...
                 CHECK(compact->utf8_length == ascii->length);
             }
             else {
-                CHECK(compact->utf8 != data);
+                CHECK(_PyUnicode_UTF8(op) != data);
             }
         }
 
-        if (compact->utf8 == NULL)
+        if (_PyUnicode_UTF8(op) == NULL)
```
Comment (colesbury) [inline Objects/unicodeobject.c:711]: This check is not safe in the free threading build: if `utf8` is not NULL, then the length is valid, but the opposite condition doesn't tell you anything.

I think we should just `#ifndef Py_GIL_DISABLED` around this.

File: Objects/unicodeobject.c [hunk 6]
```diff
@@ -114,7 +114,7 @@ NOTE: In the interpreter's initialization phase, some globals are currently
 
 static inline char* _PyUnicode_UTF8(PyObject *op)
 {
-    return (_PyCompactUnicodeObject_CAST(op)->utf8);
+    return FT_ATOMIC_LOAD_PTR_ACQUIRE(_PyCompactUnicodeObject_CAST(op)->utf8);
```
Comment (colesbury) [inline Objects/unicodeobject.c:117]: I don't love the atomic load here. It's not necessary for most of the call sites and I think it makes it harder to see the synchronization, but we can deal with that in follow up PRs.

**Review by colesbury — COMMENTED** [review]
Overall looks good, I think there's just one issue in _PyUnicode_CheckConsistency.

File: Objects/unicodeobject.c (hunk 5 above)
Comment (kumaraditya303) [reply Objects/unicodeobject.c:711]: done

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
> The buildbot AMD64 CentOS9 NoGIL Refleaks 3.x has failed when building commit https://github.com/python/cpython/commit/3c168f7f79d1da2323d35dcf88c2d3c8730e5df6.

The failure looks unrelated:

```
0:12:32 load avg: 8.03 [466/481/1] test_free_threading worker non-zero exit code (Exit code -6 (SIGABRT)) -- running (3): (...)

(...)

Races assigning to __dict__ should be thread safe ...

python: Objects/obmalloc.c:1219: process_queue: Assertion `buf->rd_idx == buf->wr_idx' failed.
Fatal Python error: Aborted

Thread 0x00007fbb557fa640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 162 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Current thread 0x00007fbb55ffb640 (most recent call first):
  Garbage-collecting
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 163 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb567fc640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 162 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb56ffd640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 158 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb577fe640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 158 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb57fff640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 162 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb16ffd640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 158 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb15ffb640 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 158 in writer_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 996 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1054 in _bootstrap_inner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1016 in _bootstrap

Thread 0x00007fbb5d092740 (most recent call first):
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/threading.py", line 1105 in join
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/test_free_threading/test_dict.py", line 188 in test_racing_set_object_dict
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/case.py", line 606 in _callTestMethod
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/case.py", line 660 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/case.py", line 716 in __call__
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 122 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 84 in __call__
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 122 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 84 in __call__
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 122 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 84 in __call__
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 122 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/suite.py", line 84 in __call__
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/unittest/runner.py", line 259 in run
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 58 in _run_suite
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 38 in run_unittest
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 136 in test_func
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/refleak.py", line 132 in runtest_refleak
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 88 in regrtest_runner
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 139 in _load_run_test
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 184 in _runtest_env_changed_exc
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 284 in _runtest
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/single.py", line 313 in run_single_test
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/worker.py", line 83 in worker_process
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/worker.py", line 118 in main
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/test/libregrtest/worker.py", line 122 in <module>
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/runpy.py", line 88 in _run_code
  File "/home/buildbot/buildarea/3.x.itamaro-centos-aws.refleak.nogil/build/Lib/runpy.py", line 198 in _run_module_as_main

Extension modules: _testcapi (total: 1)
```

**Pull request merged**